### PR TITLE
Yaml Validator - Iteration One

### PR DESCRIPTION
**Pull request by abaskk-msft**
### YAML Validator - Iteration One
- This is a initial version of the core YAML Validator with the CLI scaffolding.
- Note -> YAML is converted to JSON then validated against the JSON schema
- The YAML validator has been unit tested
- Here is a rough UML of the program structure
image

#### 1. Validation Error Handling
- For now, I catch errors but don't handle errors properly, I want some feedback so I can add that in my next iteration.
- Cases
	- Loading Invalid Yaml File
	- Empty YAML file
	- Directory with several yamls -> do we gracefully move on to the next file or do we throw and error, log to standard error and exit with an error code?
- Error codes when exiting with error (like persistence)

#### 3. Verbosity flag and less verbose error trace
- Add a flag for verbosity
- Currently, the only output is whether a yaml file is valid or not. For the sake of clarity, I've commented out the part of the code where a schema error path is given. This is immensely verbose. The current output is simply whether the YAML is valid or not.
- The intention would be to add a stack trace from this verbose output to show the exact location where someone's YAML has errored
- here is an example of the current error output (it outputs the entire json tree), we can extract a readable path from this.
- You can copy and paste the below json [here](https://jsonviewer.stack.hu/) to see how verbose it is

```
{"valid":false,"evaluationPath":"","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#","instanceLocation":"","errors":{"oneOf":"Expected 1 matching subschema but found 0"},"details":[{"valid":false,"evaluationPath":"/oneOf/0","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/oneOf/0","instanceLocation":"","details":[{"valid":false,"evaluationPath":"/oneOf/0/$ref","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef","instanceLocation":"","details":[{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$","instanceLocation":"/App","details":[{"valid":true,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Control","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/properties/Control","instanceLocation":"/App/Control","details":[{"valid":true,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Control/$ref","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/subschemas/control-type-schema.json#","instanceLocation":"/App/Control","annotations":{"title":"Microsoft Power Apps Control Type","description":"The type of the control","examples":["AddMedia","Arrow","Attachments","Audio","BarChart","Barcode","BarcodeScanner","Classic/Button","Camera","Classic/CheckBox","Circle","Classic/ComboBox","DataCard","DataGrid","DataTable","DataTableColumn","Classic/DatePicker","Classic/DropDown","Export","FluidGrid","Form","FormsPro","FormViewer","Gallery","GalleryTemplate","Group","GroupContainer","Hexagon","HtmlViewer","Classic/Icon","Image","Import","PenInput","Label","Legend","LineChart","ListBox","Lookup","Microphone","Octagon","PartialCircle","PDFViewer","Pentagon","PieChart","PowerBI","Classic/Radio","Rating","Rectangle","RichTextEditor","Classic/Slider","Star","Classic/StreamVideo","Classic/TextInput","Timer","Classic/Toggle","Triangle","TypedDataCard","Video","FluentV8/Button","FluentV8/CheckBox","FluentV8/ComboBox","FluentV8/DatePicker","FluentV8/Label","FluentV8/Radio","FluentV8/Rating","FluentV8/Slider","FluentV8/TextBox","FluentV8/Toggle","Badge","Button","CheckBox","ComboBox","DatePicker","DropDown","Header","InfoButton","Link","NumberInput","Progress","Radio","Slider","Spinner","StreamVideo","TabList","Text","TextInput","Toggle","Table","Icon","Avatar","Breadcrumb","SpinButton","Switch","PowerVirtualAgentsChat","IoTDeviceManager","AnnotateInMR","CopilotAnswer","Copilot","Chatbot","BusinessCardReader","BarcodeReader","AddressInput","ViewIn3D","ViewShapeInMR","ViewInMR","TextRecognizer","ReceiptProcessor","ObjectDetector","Map","FormProcessor","MeasureInMR","EntityForm","PcfDataField","SalesNavigatorAccount","SalesNavigatorLead","SpeechControl"]}}]},{"valid":true,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Properties","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/properties/Properties","instanceLocation":"/App/Properties","details":[{"valid":true,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Properties/$ref","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/subschemas/control-property-schema.json#","instanceLocation":"/App/Properties","annotations":{"title":"Microsoft Power Apps Properties","description":"The properties of the control","properties":[]}}]},{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/properties/Children","instanceLocation":"/App/Children","details":[{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children/items","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/properties/Children/items","instanceLocation":"/App/Children/0","details":[{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children/items/$ref","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef","instanceLocation":"/App/Children/0","details":[{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children/items/$ref/patternProperties/^[^\\\\]*$","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$","instanceLocation":"/App/Children/0/Host","errors":{"required":"Required properties [\u0022Control\u0022] are not present"},"details":[{"valid":true,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children/items/$ref/patternProperties/^[^\\\\]*$/properties/Variant","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/properties/Variant","instanceLocation":"/App/Children/0/Host/Variant","annotations":{"description":"The variant of the control"}},{"valid":false,"evaluationPath":"/oneOf/0/$ref/patternProperties/^[^\\\\]*$/properties/Children/items/$ref/patternProperties/^[^\\\\]*$/additionalProperties","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/$defs/ControlDef/patternProperties/^[^\\\\]*$/additionalProperties","instanceLocation":"/App/Children/0/Host/Contro","errors":{"":"All values fail against the false schema"}}]}]}]}]}]}]}]},{"valid":false,"evaluationPath":"/oneOf/1","schemaLocation":"file:///C:/Users/t-abaskar/Desktop/Work/PowerApps-Tooling/bin/Debug/YAMLValidator/schema/pa.yaml-schema.json#/oneOf/1","instanceLocation":"","errors":{"type":"Value is \u0022object\u0022 but should be \u0022array\u0022"}}]}


```

#### 4. Come up with and add more unit tests for resilience 
- Also clean up some of the unit testing code by moving it into a function

### CODE REVIEW

File: src/YAMLValidator/Utility.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public static string ReadFileData(string filePath)
+    {
+        try
+        {
+            using var reader = new StreamReader(filePath);
+            var yamlData = reader.ReadToEnd();
+            return yamlData;
+        }
+        catch (IOException)
```
Comment (mizrael) [inline src/YAMLValidator/Utility.cs:18]: remove the try/catch, let's have exceptions bubble up to the caller.

File: src/YAMLValidator/Utility.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+using YamlDotNet.RepresentationModel;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence;
+internal sealed class YamlValidatorUtility
+{
+    public static string ReadFileData(string filePath)
+    {
+        try
+        {
+            using var reader = new StreamReader(filePath);
```
Comment (mizrael) [inline src/YAMLValidator/Utility.cs:14]: FYI: you can use [File.ReadAll](https://learn.microsoft.com/en-us/dotnet/api/system.io.file.readalltext?view=net-8.0)

File: src/YAMLValidator/YamlLoader.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.ObjectModel;
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence;
+internal sealed class YamlLoader
```
Comment (mizrael) [inline src/YAMLValidator/YamlLoader.cs:7]: see all my comments on `SchemaLoader`

**Review by mizrael — CHANGES_REQUESTED** [review]

File: src/YAMLValidator/SchemaLoader.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public JsonSchema Schema { get; }
+    private const string _schemaFolderPath = "subschemas";
+
+    public SchemaLoader(string schemaPath)
+    {
+        Schema = LoadSchema(schemaPath);
+    }
+
+    private static JsonSchema LoadSchema(string schemaPath)
```
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply src/YAMLValidator/SchemaLoader.cs:17]: I'm getting this error when I make the class stateless and the function non static: `CA1822: Mark members as static`, 'Member Load does not access instance data" -> Its fine if I just supress it, right?

```
    public IReadOnlyDictionary<string, string> Load(string filePath, string pathType)
    {
        var loadedYaml = LoadFilePathData(filePath, pathType);
        return new ReadOnlyDictionary<string, string>(loadedYaml);
    }

```

File: src/YAMLValidator/YamlLoader.cs (hunk 3 above)
Comment (abaskk-msft) [reply src/YAMLValidator/YamlLoader.cs:7]: I'm getting this error when I make the class stateless and the function non static: `CA1822: Mark members as static`, 'Member Load does not access instance data" -> Its fine if I just supress it, right?

```
    public IReadOnlyDictionary<string, string> Load(string filePath, string pathType)
    {
        var loadedYaml = LoadFilePathData(filePath, pathType);
        return new ReadOnlyDictionary<string, string>(loadedYaml);
    }

```

File: src/YAMLValidator/Validator.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                JsonNode.Parse("{}");
+
+            var results = Schema.Evaluate(jsonData, _verbosityOptions);
+            // var output = JsonSerializer.Serialize(results, _serializerOptions);
+            // Console.WriteLine(output);
+            return results.IsValid;
+        }
+        catch (ArgumentOutOfRangeException e)
+        {
+            Console.WriteLine(e.Message);
```
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply src/YAMLValidator/Validator.cs]: Fixed

File: src/YAMLValidator/Validator.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        {
+            OutputFormat = OutputFormat.Hierarchical
+        };
+        // _serializerOptions = new JsonSerializerOptions
+        // {
+        //     Converters = { new EvaluationResultsJsonConverter() }
+        // };
+    }
+
+    public bool Validate()
```
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply src/YAMLValidator/Validator.cs]: quick update:
- I did[ fix this](https://github.com/microsoft/PowerApps-Tooling/blob/919ed6c42f75bc60cc6f1856ed66698cd28fdcba/src/YAMLValidator/Validator.cs), but I want new feedback, I encapsulated the returned results from the ```.evaluate function``` which produces a EvaluationResults object which has quite a few fields that may be too verbose. I know it goes against your advice of making the application stateless, but it was convenient for testing and outputting errors.

- Testing: since the encapsulated object also has a field indicating whether the whole schema matched or not I just use this for the updated tests

- OneOf edge case for false schemas: testing can be made more robust for false cases by actually looking at specific errors [here](https://github.com/microsoft/PowerApps-Tooling/blob/919ed6c42f75bc60cc6f1856ed66698cd28fdcba/src/YAMLValidator/YamlValidatorResults.cs) and validating them in tests instead of just looking at schema true or false. I want to do this in another PR though and keep the focus on this one on simple functionality testing of the validation. thoughts?

File: .version/PipelineAssemblyInfo.cs [hunk 7]
```diff
@@ -3,6 +3,9 @@
 
 // This is pipeline generated file. Do not modify. This will be replaced with the actual versions in the actual Pipeline.
 using System.Reflection;
+using System.Runtime.CompilerServices;
 [assembly: AssemblyVersion("0.0.0.0")]
 [assembly: AssemblyFileVersion("0.0.0.0")]
 [assembly: AssemblyInformationalVersion("0.0.0.0-dev-00000000")]
+[assembly: InternalsVisibleTo("YAMLValidator.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
```
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply .version/PipelineAssemblyInfo.cs:10]: TBD, did a bandaid fix because I tried to mimicking persistence and didn't work

[the remark this answers is not in the document]
Comment (mizrael) [reply .version/PipelineAssemblyInfo.cs:10]: what was the issue?

File: src/YAMLValidator/FilePathData.cs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+
+namespace Microsoft.PowerPlatform.PowerApps.Persistence;
+internal readonly record struct FilePathData(string FilePath, string SchemaPath, string FilePathType);
```
Comment (mizrael) [inline src/YAMLValidator/FilePathData.cs:6]: nit: consider renaming this to something like `ValidationRequest`

File: src/YAMLValidator/Validator.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        // to do: add verbosity flag and allow users to choose verbosity of evaluation
+        _verbosityOptions = options;
+        _serializerOptions = resultSerializeOptions;
+
+    }
+
+    public YamlValidatorResults Validate(JsonSchema schema, string yamlFileData)
+    {
+        var yamlStream = YamlValidatorUtility.MakeYamlStream(yamlFileData);
```
Comment (mizrael) [inline src/YAMLValidator/Validator.cs:27]: if `YamlStream` implements `IDisposable`, please add a `using` statement here

File: .version/PipelineAssemblyInfo.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply .version/PipelineAssemblyInfo.cs:10]:
```
  <ItemGroup>
    <InternalsVisibleTo Include="YamlValidator.Tests" Key="0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9" />
  </ItemGroup>

```
I tried adding in this flag in the .csproj but visual studio's static analysis still indicated that the YamlValidator.Tests can't access internal classes. I'm looking into this right now.

File: src/YAMLValidator/Validator.cs (hunk 9 above)
Comment (abaskk-msft) [reply src/YAMLValidator/Validator.cs:27]: [It doesn't implement the IDisposable interface from the source code it seems](https://github.com/aaubry/YamlDotNet/blob/master/YamlDotNet/RepresentationModel/YamlStream.cs.

File: .version/PipelineAssemblyInfo.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (abaskk-msft) [reply .version/PipelineAssemblyInfo.cs:10]: Note for future: changed all classes to public for now, to keep focus on features

[the remark this answers is not in the document]
Comment (mizrael) [reply .version/PipelineAssemblyInfo.cs:10]: now that classes are public, can we remove this?

File: src/YAMLValidator.Tests/ValidatorTest.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Json.Schema;
+using Microsoft.PowerPlatform.PowerApps.Persistence;
+
+namespace YamlValidatorTests;
+
+[TestClass]
```
Comment (mizrael) [inline src/YAMLValidator.Tests/ValidatorTest.cs:9]: nit: if some tests are asserting similar things, you might consider grouping them into one and parametrizing them: https://learn.microsoft.com/en-us/visualstudio/test/how-to-create-a-data-driven-unit-test?view=vs-2022

**Review by mizrael — APPROVED** [review]

**Review by petrochuk — APPROVED** [review]
